src/absil/ilwrite.fs

```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
+        try
+            reportTime showTimes "Signing Image"
+            use closeS = { new System.IDisposable with member x.Dispose() = try s.Close() with _ -> () }
+            s.SignFile filePath
         with e -> 
             failwith ("Warning: A call to StrongNameSignatureGeneration failed ("+e.Message+")")
-            (try s.Close() with _ -> ())
-            (try FileSystem.FileDelete outfile with _ -> ()) 
-            ()
-
-    reportTime showTimes "Signing Image"
-    //Finished writing and signing the binary and debug info...
+        reportTime showTimes "Signed Image"
+
+    let serializeToFile extension f emitter =
+        match emitter with
+        | EmittedFile path ->
+            tryDeleteFile path
+            f path
+        | EmittedStream stream ->
+            let tempFilePath = Path.ChangeExtension(FileSystem.GetTempFilePathShim(), extension)
+            tryDeleteFile tempFilePath
+            f tempFilePath
+            use tempFileStream = FileSystem.FileStreamReadShim(tempFilePath)
+            tempFileStream.CopyTo(stream)
+
+    let serializeToPdb = serializeToFile ".pdb"
+    let serializeToMdb = serializeToFile ".mdb"
+
+    let emitOrFail (e: EmitStreamProvider) = 
+        match e.Value with
+        | Success stream -> stream
+        | Failure failure -> failwith failure
+
+    use outfileStream = new MemoryStream()
+    let os =  new BinaryWriter(outfileStream)
+
+    let pdbData,debugDirectoryChunk,debugDataChunk,textV2P,mappings =
+        gen os (pdbP |> Option.isSome)
+
+    os.Flush()
+    outfileStream.Position <- 0L
+
+    dumpDebugInfo |> Option.map emitOrFail |> Option.iter (writeDumpDebugInfo pdbData)
+
+    match outfileP |> emitOrFail, pdbP |> Option.map emitOrFail, mdbP |> Option.map emitOrFail, signer with
+    | EmittedStream(s), None, None, None ->
+        outfileStream.CopyTo(s)
+    | EmittedStream(s), pdb, mdb, signer ->
+        // both pdb, mdb and signer implementation require the output file, so let's create it
+        let path = FileSystem.GetTempFilePathShim()
+        using (FileSystem.FileStreamCreateShim(path)) outfileStream.CopyTo
+
+        pdb |> Option.iter (serializeToPdb (writePdb (pdbData, textV2P, debugDirectoryChunk, debugDataChunk) path))
+        mdb |> Option.iter (serializeToMdb (WriteMdbInfo pdbData path))
+        signer |> Option.iter (signTo path)
+
+        use oStream  = FileSystem.FileStreamReadShim(path)
+        oStream.CopyTo(s)
+    | EmittedFile(path), pdb, mdb, signer ->
+        using (FileSystem.FileStreamCreateShim(path)) outfileStream.CopyTo
+        path |> FileSystemUtilites.setExecutablePermission 
+
+        pdb |> Option.iter (serializeToPdb (writePdb (pdbData, textV2P, debugDirectoryChunk, debugDataChunk) path))
+        mdb |> Option.iter (serializeToMdb (WriteMdbInfo pdbData path))
+        signer |> Option.iter (signTo path)
 
     mappings
 
 
+let private optionOfLazy (l: Lazy<'T>) =
```

Comment:
i'll defend the use of Lazy in a separate comment because i think is usefull and is a strong contract.
But i can inline (as inline refactoring) this function so less confusion ( + a comment where used about the clean used files behaviour )